File: 116__Populating_Next_Right_Pointers_in_Each_Node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):

        if root is None:
            return None

        cur_level = []
        cur_level.append(root)

        while cur_level:
            next_level = []
            for i in range(len(cur_level)):
                cur_node = cur_level[i]
                if i+1 < len(cur_level):
                    cur_node.next = cur_level[i+1]
                else:
                    cur_node.next = None
                if cur_node.left: next_level.append(cur_node.left)
                if cur_node.right: next_level.append(cur_node.right)
            cur_level = next_level

File: Java/116__Populating_Next_Right_Pointers_in_Each_Node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val,Node _left,Node _right,Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/
class Solution {
    public Node connect(Node root) {
        if (root == null) return root;
        Queue<Node> visited = new LinkedList<>();
        visited.add(root);
        int levelNodesCount = visited.size();
        while (!visited.isEmpty()) {
            Node node = visited.poll();
            if (levelNodesCount > 1) node.next = visited.peek();
            if (node.left != null) visited.add(node.left);
            if (node.right != null) visited.add(node.right);
            levelNodesCount--;
            if (levelNodesCount == 0)levelNodesCount = visited.size();
        }
        return root;
    }
}
